src/IO/LzmaReadBuffer.cpp

```diff
@@ -0,0 +1,81 @@
+#include <IO/LzmaReadBuffer.h>
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int LZMA_STREAM_DECODER_FAILED;
+}
+LzmaReadBuffer::LzmaReadBuffer(std::unique_ptr<ReadBuffer> in_, size_t buf_size, char * existing_memory, size_t alignment)
+    : BufferWithOwnMemory<ReadBuffer>(buf_size, existing_memory, alignment), in(std::move(in_)), eof(false)
+{
+    lstr = LZMA_STREAM_INIT;
+    lstr.allocator = nullptr;
+    lstr.next_in = nullptr;
+    lstr.avail_in = 0;
+    lstr.next_out = nullptr;
+    lstr.avail_out = 0;
+
+    // 500 mb
+    uint64_t memlimit = 500 << 20;
+
+    lzma_ret ret = lzma_stream_decoder(&lstr, memlimit, LZMA_CONCATENATED);
+    // lzma does not provide api for converting error code to string unlike zlib
+    if (ret != LZMA_OK)
+        throw Exception(
```

Comment:
There is fmt-style constuctor:

`Exception(ErrorCodes::LZMA_STREAM_DECODER_FAILED, "template {} {}", var1, var2)`